src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,713 @@
+// src/main/c/NativeModule.cpp - native funcs for Scala NativeModule
+#include "hail/NativeModule.h"
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <assert.h>
+#include <jni.h>
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <unistd.h>
+#include <atomic>
+#include <memory>
+#include <mutex>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#if 0
+#define D(fmt, ...) { \
+  char buf[1024]; \
+  sprintf(buf, fmt, ##__VA_ARGS__); \
+  fprintf(stderr, "DEBUG: %s,%d: %s", __FILE__, __LINE__, buf); \
+}
+#else
+#define D(fmt, ...) { }
+#endif
+
+namespace hail {
+
+namespace {
+
+// File-polling interval in usecs
+const int kFilePollMicrosecs = 50000;
+
+// Timeout for compile/link of a DLL
+const int kBuildTimeoutSecs = 120;
+
+// A quick-and-dirty way to get a hash of two strings, take 80bits,
+// and produce a 20byte string of hex digits.
+std::string hash_two_strings(const std::string& a, const std::string& b) {
+  auto hashA = std::hash<std::string>()(a);
+  auto hashB = std::hash<std::string>()(b);
+  hashA ^= (0x3ac5*hashB);
+  hashB &= 0xffff;
```

Comment:
No, those are 64bit values (I think size_t) - and we're *multiplying* hashB by that 16bit value
to get a 64bit value to mix with hashA.  Then we're also taking 16bits of hashB to bring it
up to 80bits in total.- chosen as a compromise between being long enough to have a very
low probability of collisions between any pair of modules ever generated by a single user
(10 per second for 50 years would be about 2**34), and giving hex-encoded filenames which
aren't crazy long.

I'm relying on std::hash<std::string> to give unlikely-to-collide values, but since that is its
main purpose it seems like a good place to start.  And if it proves false then
it's not much work to plug in a stronger hash.

More realistically, at some point we'll probably want to have a strategy for deleting very-old-and-not-currently-in-use modules, so that a user doesn't accumulate more than a few hundred.
But that needs a bit of extra care about to guarantee the "not-currently-in-use" condition.